ENH: Add FSL `eddy_quad` interface

richford opened the pull request:
<!--

Pull-request guidelines
-----------------------

1. If you would like to list yourself as a Nipype contributor and your name is not mentioned please modify .zenodo.json file.
2. By submitting this request you acknowledge that your contributions are available under the Apache 2 license.
3. Use a descriptive prefix for your PR: ENH (enhancement), FIX, TST, DOC, STY, REF (refactor), WIP (Work in progress)
4. Run `make check-before-commit` before submitting the PR.

-->
## Summary
<!-- Please reference any related issue and use fixes/close to automatically
close them, if pertinent -->

This PR adds an interface to `eddy_quad`, one of two tools in the EDDY QC framework.
See the [EDDY QC documentation](https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/eddyqc/UsersGuide) for further details. Thanks to @akeshavan for help with my first nipype PR.

## List of changes proposed in this PR (pull-request)
<!-- We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

- Add `EddyQuadInputSpec` class
- Add `EddyQuadOutputSpec` class
- Add `EddyQuad` class with doctest

## Acknowledgment

- [x] \(Mandatory\) I acknowledge that this contribution will be available under the Apache 2 license.

richford commented:
Hmm, the travis PR checks failed, but only for Python 3.4 because `xdist` could not be loaded. Not sure why this is happening. I don't think that anything I changed in the PR is causing this. Any help would be appreciated.

richford commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,220 @@ def _run_interface(self, runtime):
+        argstr='--field=%s',
+        desc="TOPUP estimated field (in Hz)",
+    )
+    slspec = File(
```
Thanks. Resolved in [8bac064](https://github.com/nipy/nipype/pull/2825/commits/8bac0647eb1d324eaa8f34005e4e9dc1de6c4e05).

richford commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,220 @@ def _run_interface(self, runtime):
+    )
+    verbose = traits.Bool(
+        False,
+        mandatory=False,
```
Thanks. Resolved in [8bac064](https://github.com/nipy/nipype/pull/2825/commits/8bac0647eb1d324eaa8f34005e4e9dc1de6c4e05).

richford commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,220 @@ def _run_interface(self, runtime):
+        desc="TOPUP estimated field (in Hz)",
+    )
+    slspec = File(
+        mandatory=False,
```
Thanks. Resolved in [8bac064](https://github.com/nipy/nipype/pull/2825/commits/8bac0647eb1d324eaa8f34005e4e9dc1de6c4e05).

richford commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,220 @@ def _run_interface(self, runtime):
+        desc="Output directory - default = '<base_name>.qc'",
+    )
+    field = File(
+        mandatory=False,
```
Thanks. Resolved in [8bac064](https://github.com/nipy/nipype/pull/2825/commits/8bac0647eb1d324eaa8f34005e4e9dc1de6c4e05).

richford commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,220 @@ def _run_interface(self, runtime):
+        mandatory=True,
+        usedefault=True,
+        argstr='--output-dir=%s',
+        desc="Output directory - default = '<base_name>.qc'",
```
Thanks. Incorporated your changes in [8bac064](https://github.com/nipy/nipype/pull/2825/commits/8bac0647eb1d324eaa8f34005e4e9dc1de6c4e05).

richford commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,220 @@ def _run_interface(self, runtime):
+
+class EddyQuadInputSpec(FSLCommandInputSpec):
+    base_name = traits.Str(
+        'eddy_corrected',
```
Thanks. Resolved in [8bac064](https://github.com/nipy/nipype/pull/2825/commits/8bac0647eb1d324eaa8f34005e4e9dc1de6c4e05).

richford commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,220 @@ def _run_interface(self, runtime):
+    )
+    bvec_file = File(
+        exists=True,
+        mandatory=False,
```
Thanks. Resolved in [8bac064](https://github.com/nipy/nipype/pull/2825/commits/8bac0647eb1d324eaa8f34005e4e9dc1de6c4e05).

richford commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,220 @@ def _run_interface(self, runtime):
+    base_name = traits.Str(
+        'eddy_corrected',
+        argstr='%s',
+        desc="Basename (including path) specified when running EDDY",
```
I improved the description in [8bac064](https://github.com/nipy/nipype/pull/2825/commits/8bac0647eb1d324eaa8f34005e4e9dc1de6c4e05). I think it should be clear to EDDY users. But honestly, I'm no EDDY expert myself.

richford commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,220 @@ def _run_interface(self, runtime):
+        )
+
+        outputs['out_avg_b0_png'] = glob(os.path.abspath(
+            os.path.join(out_dir, 'avg_b0_pe*.png')
```
Good call. In [8bac064](https://github.com/nipy/nipype/pull/2825/commits/8bac0647eb1d324eaa8f34005e4e9dc1de6c4e05), I now inspect the `qc.json` which has flags that indicate the presence of certain output files and the number of expected files for each type.

richford commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,220 @@ def _run_interface(self, runtime):
+            os.path.join(out_dir, 'vdm.png')
+        )
+
+        if os.path.exists(vdm):
```
Thanks. In [8bac064](https://github.com/nipy/nipype/pull/2825/commits/8bac0647eb1d324eaa8f34005e4e9dc1de6c4e05), I inspect the `qc.json` file to see if those files should exist.

richford commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,220 @@ def _run_interface(self, runtime):
+    def _list_outputs(self):
+        from glob import glob
+        outputs = self.output_spec().get()
+        out_dir = self.inputs.output_dir
```
Thanks. Resolved in [8bac064](https://github.com/nipy/nipype/pull/2825/commits/8bac0647eb1d324eaa8f34005e4e9dc1de6c4e05).

richford commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,220 @@ def _run_interface(self, runtime):
+    output_spec = EddyQuadOutputSpec
+
+    def __init__(self, **inputs):
+        super(EddyQuad, self).__init__(**inputs)
```
Thanks. Resolved in [8bac064](https://github.com/nipy/nipype/pull/2825/commits/8bac0647eb1d324eaa8f34005e4e9dc1de6c4e05).

richford commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,220 @@ def _run_interface(self, runtime):
+              "file is present")
+    )
+    output_dir = traits.Str(
+        'eddy_corrected.qc',
```
Thanks. I think I resolved this in [8bac064](https://github.com/nipy/nipype/pull/2825/commits/8bac0647eb1d324eaa8f34005e4e9dc1de6c4e05). Is that correct?

richford commented:
Thanks for your review @effigies. I responded inline to your reviews and closed the ones that I felt were trivially resolved, leaving the more involved ones to you to review and close if you agree that I've addressed them. Thanks again for your help!

effigies commented on .zenodo.json:
```diff
@@ -589,7 +589,12 @@
+      "affiliation": "University of Washington",
+      "name": "Richie-Halford, Adam",
+      "orcid": "0000-0001-9276-9084"
+    },
```
```suggestion
    }
```

effigies commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,206 @@ def _run_interface(self, runtime):
+class EddyQuadOutputSpec(TraitedSpec):
+    out_qc_json = File(
+        exists=True,
+        mandatory=True,
```
```suggestion
```

effigies commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,206 @@ def _run_interface(self, runtime):
+
+    out_qc_pdf = File(
+        exists=True,
+        mandatory=True,
```
```suggestion
```

effigies commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,206 @@ def _run_interface(self, runtime):
+    )
+
+    out_avg_b_png = traits.List(
+        File(
```
```suggestion
        File(exists=True),
```

effigies commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,206 @@ def _run_interface(self, runtime):
+            exists=True,
+            mandatory=True,
+            desc=("Image showing mid-sagittal, -coronal and -axial slices of "
+                  "each averaged b-shell volume.")
```
`desc` should be a parameter to `List`, not `File`, and you can drop `mandatory`. I've suggested removing it from the other output spec files... I don't think it's enforced, but if it is, it will cause problems if you have the `remove_unnecessary_outputs` config option set, and no node takes those files as inputs.

effigies commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,206 @@ def _run_interface(self, runtime):
+    )
+
+    out_avg_b0_pe_png = traits.List(
+        File(
```
```suggestion
        File(exists=True),
```

effigies commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,206 @@ def _run_interface(self, runtime):
+            mandatory=False,
+            desc=("Image showing mid-sagittal, -coronal and -axial slices of "
+                  "each averaged pe-direction b0 volume. Generated when using "
+                  "the -f option.")
```
Again.

effigies commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,206 @@ def _run_interface(self, runtime):
+    )
+
+    out_cnr_png = traits.List(
+        File(
```
```suggestion
        File(exists=True),
```

effigies commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,206 @@ def _run_interface(self, runtime):
+            mandatory=False,
+            desc=("Image showing mid-sagittal, -coronal and -axial slices of "
+                  "each b-shell CNR volume. Generated when CNR maps are "
+                  "available.")
```
Again.

effigies commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,206 @@ def _run_interface(self, runtime):
+
+    out_vdm_png = File(
+        exists=True,
+        mandatory=False,
```
```suggestion
```

effigies commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,206 @@ def _run_interface(self, runtime):
+
+    out_residuals = File(
+        exists=True,
+        mandatory=False,
```
```suggestion
```

effigies commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,206 @@ def _run_interface(self, runtime):
+
+    out_clean_volumes = File(
+        exists=True,
+        mandatory=False,
```
```suggestion
```

effigies commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,206 @@ def _run_interface(self, runtime):
+        outputs['out_qc_pdf'] = os.path.join(out_dir, 'qc.pdf')
+
+        with open(outputs['out_qc_json']) as fp:
+            qc = json.load(fp)
```
Ah, we'd prefer `glob` to this. The reason is that `_list_outputs()` should be runnable without having already run the interface. With `glob` you'll just get an empty list, while attempting to read a JSON file will actually fail.

If there's no way to predict the outputs entirely without running, that's fine. I was just suggesting that, if you could, that's a little nicer. But we should only depend on `self.inputs` to determine the expected outputs. And I can't remember whether you used `sorted(glob(...))` before, but I would recommend sorting the list so it's consistent across runs. (`glob` does not have to return sorted lists.)

effigies commented:
Hi, just a reminder that the 1.1.8 release is targeted for January 28. Please let us know if you'd like to try to finish this up for that release.

richford commented:
@effigies I think this is ready for review. I changed back to `glob` and tried to rely on `self.inputs` as much as I could. I also added a fallback check for the `output_dir` after @akeshavan and I found that `self.inputs.output_dir` was undefined when it used the default even though it passed the correct value to the FSL CLI.

richford commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,199 @@ def _run_interface(self, runtime):
+
+
+class EddyQuadOutputSpec(TraitedSpec):
+    out_qc_json = File(
```
Good call. Fixed in [2fe91b9](https://github.com/nipy/nipype/pull/2825/commits/2fe91b9d957b92e8042bb68ee29ecbcca34b8b81)

richford commented on nipype/interfaces/fsl/epi.py:
```diff
@@ -1232,3 +1232,199 @@ def _run_interface(self, runtime):
+        # If the output directory isn't defined, the interface seems to use
+        # the default but not set its value in `self.inputs.output_dir`
+        if not isdefined(self.inputs.output_dir):
+            out_dir = os.path.abspath(os.path.basename(self.inputs.base_name) + '.qc.nii.gz')
```
@akeshavan, we tested this on your system. I think this is correct but **it is** very weird. Could you confirm? (I don't want to take the hours needed to build the docker image on my laptop. :smirk:)

effigies commented:
Oh, can you please merge `master` to fix the tests?

effigies commented:
@akeshavan Feel free to merge if you're happy with this.

richford commented:
> Oh, can you please merge `master` to fix the tests?

Just rebased onto master (with a force push for the change in history).

effigies commented:
I think we can take your word on the default `output_dir`. Please open a new PR if you discover it should be something else.
